### PR TITLE
Add Warning for Deprecated Modules in `Init` (Second Revision)

### PR DESCRIPTION
**Pull request by Maed223**
[Jira Ticket](https://hashicorp.atlassian.net/browse/TF-12601?atlOrigin=eyJpIjoiZGJlMWU1MzZiMWQ1NGRjNjkzNTQzOTY1MjI4NjA4YTQiLCJwIjoiaiJ9)

## Target Release

1.9.x

### ENHANCEMENTS

- `terraform init`:  Displays warnings when modules marked as deprecated are in use.

## Description

Module deprecations warnings are surfaced upon running `terraform init`. This includes both cases of where the module does, or does not, need installation. Furthermore, the warning can be raised for both modules that are directly used in configuration, as well as any external dependencies they depend on.

Screenshot 2024-05-23 at 12 23 01 PM

### CODE REVIEW

File: internal/initwd/module_install.go
Comment (Maed223) [inline internal/initwd/module_install.go:302]:
```suggestion
											var additionalInfo []string
											if modVersion.Deprecation.Reason != "" {
												additionalInfo = append(additionalInfo, modVersion.Deprecation.Reason)
											}
											if modVersion.Deprecation.Link != "" {
												additionalInfo = append(additionalInfo, fmt.Sprintf("More information: %s", modVersion.Deprecation.Link))
											}
											var detail string
											if len(additionalInfo) == 1 {
												detail = additionalInfo[0]
											} else {
												detail = strings.Join(additionalInfo, "\n\n")
											}
```

Could use some suggestion on this bit, as it is repeated elsewhere in `installRegistryModules`. First thought is have this in a separate function, but that seems out of place in this file

File: internal/initwd/module_install.go
Comment (Maed223) [reply internal/initwd/module_install.go:65]: Not exactly. This is used to only expose the Extra field if it's been given the IsPublic function.

In _internal/tfdiags/diagnostic.go_
```go
// PublicExtraInfo is an interface for marking Extra field that contain public extra information
type PublicExtraInfo interface {
	IsPublic()
}
```

In _internal/command/views/json/diagnostic.go_
```go
        extra := diag.ExtraInfo()
	var publicExtra tfdiags.PublicExtraInfo
	if e, ok := extra.(tfdiags.PublicExtraInfo); ok {
		publicExtra = e
	}

	diagnostic := &Diagnostic{
		Severity: sev,
		Summary:  desc.Summary,
		Detail:   desc.Detail,
		Address:  desc.Address,
		Extra:    publicExtra,
	}
```

Otherwise we'd be needlessly showing these Extra fields when they aren't meant for display.

File: internal/initwd/module_install_test.go
Comment (Maed223) [reply internal/initwd/module_install_test.go]: Included 👍  https://github.com/hashicorp/terraform/pull/35231/commits/9b6fc3b1fdcbffb6939bcbc832e2f5c7f3a145bc

File: internal/initwd/module_install_test.go
Comment (Maed223) [reply internal/initwd/module_install_test.go]: Included 👍  https://github.com/hashicorp/terraform/pull/35231/commits/9b6fc3b1fdcbffb6939bcbc832e2f5c7f3a145bc

**Comment by Maed223** [comment]
@uturunku1 

> * When there is multiple diagnostic warnings (that include deprecation and non-deprecation warnings), all the deprecation warnings show up intertwined with non-deprecation warnings, like: deprecation diagnostic, then non-deprecation diagnostic, another deprecation diagnostic, another non-deprecation diagnostic, etc. I kind of get why that's happening. But I wonder if there is any way that we could line up all the deprecation diagnostics, one after the other, rather than having them all intertwined with all the other type of diagnostics?
> * Just want to double that with this new approach there is no way for the diagnostic title of a deprecation to say that a module is an indirect module. So, `Warning: Module version 1.0.1 of rds (dependency of "nested") is deprecated` is impossible, right?

**Adjusting the ordering of Diagnostics**
I'm not completely sure how the ordering is derived, though it seems to me to be that the most recently raised Diag will be the first to appear. There might be a mechanism we can introduce to Diagnostics to group them based on the `MessageCode` given in the Extra field. I'll try my hand at that and see if it's feasible.

**Denoting Nested Dependency Deprecations**
With this approach we won't be able to denote that they are nested dependencies. Specifically this is because we're adding the diagnostics in the module walker function, and it has no way (without more extensive changes) to know where a given module sits in the hierarchy.

File: internal/initwd/module_install.go
Comment (Maed223) [reply internal/initwd/module_install.go]: Sounds good, included 👍 https://github.com/hashicorp/terraform/pull/35231/commits/cacc5dffa2a2aeeb2906c337384bebf8ee0da0e8

File: internal/initwd/module_install.go
Comment (brandonc) [inline internal/initwd/module_install.go:56]: Once this is merged, fields can't be modified or removed. Can you add some godoc comments about this being public API?

File: internal/initwd/module_install.go
Comment (brandonc) [inline internal/initwd/module_install.go:281]: I don't think Namespace is enough to identify this module. Was thinking `addr.Package.String()`

Comment (brandonc) [reply internal/initwd/module_install.go:281]: I like that errors here will not impede init from taking place. But also this feels like a higher level of log than DEBUG. Can we promote this to [WARN] ?

File: internal/initwd/module_install.go
Comment (brandonc) [inline internal/initwd/module_install.go:1004]: The name of the module... For example, "foo" in `module "foo" {` is not the thing that is specifically deprecated, and wording it this way is slightly confusing because of the word "of"... plus, it prevents default consolidation from taking place.

If your config says this:

```
module "A" {
  source = "app.staging.terraform.io/bcroft/example-module/aws"
  version = "1.0.0"
}
```

Should the message say "Module version 1.0.0 of A is deprecated" or should it say "Module version 1.0.0 of app.staging.terraform.io/bcroft/example-module/aws is deprecated" ?

The latter seems much more accurate, plus all the usage of that particular module will be consolidated by the default consolidator. Otherwise you get all the warnings in each usage since the summary varies by name

**Comment by brandonc** [comment]
@uturunku1 @Maed223 Even though we can't be explicit about the name of a module that contains a deprecation, the source code location that was added gives a hint:

╷
│ Warning: Module version 1.0.0 of A is deprecated
│
│   on .terraform/modules/example-module/main.tf line 1:
│    1: module "A" {
│
╵

This warning shows that module A is a dependency of "example-module"

**Comment by brandonc** [comment]
@uturunku1 @Maed223 As far as the interlaced warnings go, they ultimately end up being grouped by the module in question. You may see a deprecation warning for a module followed by the warnings generated by that specific module. Seems fine.

File: internal/initwd/module_install.go
Comment (Maed223) [reply internal/initwd/module_install.go:61]: Include 👍  https://github.com/hashicorp/terraform/pull/35231/commits/838f23b5f94dccbbcd8fc9a8dedacc294b72944c

File: internal/initwd/module_install.go
Comment (Maed223) [reply internal/initwd/module_install.go]: Good point now that I think about it– included 👍  https://github.com/hashicorp/terraform/pull/35231/commits/e3fc6d2d30c24cd5ceff57d8b3bb6fb23cadebe5

File: internal/initwd/module_install.go
Comment (Maed223) [reply internal/initwd/module_install.go]: included 💯   https://github.com/hashicorp/terraform/pull/35231/commits/e3fc6d2d30c24cd5ceff57d8b3bb6fb23cadebe5

File: internal/initwd/module_install.go
Comment (brandonc) [inline internal/initwd/module_install.go:60]: Don't mention TFC or TFC agent -- also "correcsponding"

```suggestion
// about the deprecation of a module version. This ends up being serialized as extra data within a
// diagnostic and should be considered public API.
```

File: internal/initwd/module_install.go
Comment (brandonc) [inline internal/initwd/module_install.go:287]: You may have missed my previous feedback about the `Namespace` of the module not being enough information?

File: internal/initwd/module_install.go
Comment (Maed223) [reply internal/initwd/module_install.go]: Made an edit to the commit https://github.com/hashicorp/terraform/pull/35231/commits/1368c93862d046729782fc91107ce683191be0c3

File: internal/initwd/module_install.go
Comment (Maed223) [reply internal/initwd/module_install.go]: Oh whoops! Think my brain squashed your suggestion with the note on how to include it in the diag itself.

Made an edit to the commit to include https://github.com/hashicorp/terraform/pull/35231/commits/7dd73600076f942118a48a8fbe215aee515708ee

**Review by brandonc — APPROVED** [review]
OK I am good with the changes but get another approver, too

File: internal/registry/response/module_versions.go
Comment (nfagerlund) [inline internal/registry/response/module_versions.go:26]: CJ and I pondered for a bit during review whether we needed an explicit test case to ensure _registries that do not support deprecation_ continued to work as expected. But I think I've decided that that's fine, because the absence of the deprecation field in the deserialized json is coherent in both cases — that is, a module from a registry that doesn't support deprecation is always by definition _not deprecated,_ so it's always coherent to treat it as not deprecated. 👍🏼

**Review by nfagerlund — CHANGES_REQUESTED** [review]
I'm requesting changes on some very small problems, but other than that, I think this is in very good shape and would be happy to merge it once those are fixed! 

To make sure we're aligned on intent here, this is my understanding of the core logic and its differences from the previous effort: 

- We're emitting deprecation warning diagnostics _as we go_ during module installation.
- The existing automatic diagnostic consolidation logic works by looking for identical summaries, but deprecation warnings for different modules have _different summary text,_ so we won't get entire deprecated modules hidden behind a "similar warnings" message. 👍🏼 (Conversely, if you got a ton of warnings for _the same module,_ yeah you probably do want those compressed.) 
- The deprecation warnings will be largely grouped together in the UI, just because of chronology of when they're emitted. 👍🏼
- TFC will be receiving individual diagnostics, and gets to decide on its own how to group or present them. The included public ExtraInfo with structured data about the deprecation will enable that!
- The optional "public extra info" interface prevents unwanted existing ExtraInfo types from leaking across the public diags API.

**Review by nfagerlund — APPROVED** [review]
Approved!

**Pull request merged**
